README.md
# franka3

franka3 is an experimental cipher that can be used to encrypt file content. Please do not use it for actual use case but feel free to try it. Make a backup copy of your original file if you plan to use franka3 to encrypt it.


changelog.txt
# Changelog
All notable changes to this project will be documented in this file.
The date is follow when it is published to github and not the date the file created

## [0.1.3] - 20-SEP-2022
### Added
-Introduce initial substitution gate
-Add option
-Support for more characters

### Changed
-Limit 16kb buffer to 2kb for file encryption in option 1

## [0.1.2] - 07-SEP-2021
### Added
-Add check for initial phase and decipher phase

### Fixed
-Fix max value of each phase still at 26. After this fix, max value is now 33

## [0.1.1] - 06-SEP-2021
### Added
-Support for more characters such as /,?,\a,\b,\e,\f,newline,\r,\t,\v,'

### Fixed
-Decrypting content with wrong phase won't cause program to hang

## [0.1.0] - 15-AUG-2021
### Added
-Initial release

main.h
// Date created: 20-SEP-2021 MONDAY
// This source file is part of the franka3 open source project
// Copyright (c) 2021 nicklaus yap ken yik
// version: 0.1.3

struct franka3{
int phase1;
int phase2;
int phase3;
};

// initialSubGate 1
#define _1a 5
#define _1b 11
#define _1c 13
#define _1d 6
#define _1e 12
#define _1f 7
#define _1g 4
#define _1h 17
#define _1i 22
#define _1j 26
#define _1k 14
#define _1l 20
#define _1m 15
#define _1n 23
#define _1o 25
#define _1p 8
#define _1q 24
#define _1r 21
#define _1s 19
#define _1t 16
#define _1u 1
#define _1v 9
#define _1w 2
#define _1x 18
#define _1y 3
#define _1z 10
#define _1A 32
#define _1B 38
#define _1C 40
#define _1D 33
#define _1E 39
#define _1F 31
#define _1G 44
#define _1H 49
#define _1I 53
#define _1J 41
#define _1K 47
#define _1L 42
#define _1M 50
#define _1N 52
#define _1O 35
#define _1P 51
#define _1Q 48
#define _1R 46
#define _1S 43
#define _1T 28
#define _1U 36
#define _1V 29
#define _1W 45
#define _1X 30
#define _1Y 37
#define _1Z 59
#define _1AA 65
#define _1AB 67
#define _1AC 60
#define _1AD 66
#define _1AE 58
#define _1AF 71
#define _1AG 76
#define _1AH 80
#define _1AI 68
#define _1AJ 74
#define _1AK 69
#define _1AL 77
#define _1AM 79
#define _1AN 62
#define _1AO 78
#define _1AP 75
#define _1AQ 73
#define _1AR 70
#define _1AS 55
#define _1AT 63
#define _1AU 56
#define _1AV 72
#define _1AW 57
#define _1AX 64
#define _1AY 86
#define _1AZ 92
#define _1BA 94
#define _1BB 87
#define _1BC 93
#define _1BD 85
#define _1BE 98
#define _1BF 103
#define _1BG 107
#define _1BH 95
#define _1BI 101
#define _1BJ 96
#define _1BK 104
#define _1BL 106
#define _1BM 89
#define _1BN 105
#define _1BO 102
#define _1BP 100
#define _1BQ 97
#define _1BR 82
#define _1BS 90
#define _1BT 83
#define _1BU 99
#define _1BV 84
#define _1BW 91


// extra for now
/*
#define i 
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
#define i
*/

function.c
// Date created: 14-AUG-2021 SATURDAY
// This source file is part of the franka3 open source project
// Copyright (c) 2021 nicklaus yap ken yik
// version: 0.1.3

#include <stdlib.h> // for abs(), exit()
#include "function.h"

void add(void)
{
	phase1 = phase1 + 1;
	//printf("phase1: %d\n", phase1); // debug only
	if(phase1 > 33){
		phase2 = phase2 + 1;
		//printf("phase2: %d\n", phase2); // debug only
		phase1 = 1;
	}
	else
	{
	
	}

	if(phase2 > 33){
		phase3 = phase3 + 1;
		phase2 = 1;
	}
	else
	{

	}

	if(phase3 > 33){
		phase3 = 1;
	}
	else
	{

	}
 
}

// this is the reversal of the initial sub gate option 1
void reverseInitialSubGate(int *val, int len)
{
	int i = 0;
	while (i < len) {
		// if _1a then reverse to A
		if (val[i] == 5) {
			val[i] = 'A';
			i++;
		}
		else if (val[i] == 11) {
			val[i] = 'B';
			i++;
		}
		else if (val[i] == 13) {
			val[i] = 'C';
			i++;
		}
		else if (val[i] == 6) {
			val[i] = 'D';
			i++;
		}
		else if (val[i] == 12) {
			val[i] = 'E';
			i++;
		}
		else if (val[i] == 7) {
			val[i] = 'F';
			i++;
		}
		else if (val[i] == 4) {
			val[i] = 'G';
			i++;
		}
		else if (val[i] == 17) {
			val[i] = 'H';
			i++;
		}
		else if (val[i] == 22) {
			val[i] = 'I';
			i++;
		}
		else if (val[i] == 26) {
			val[i] = 'J';
			i++;
		}
		else if (val[i] == 14) {
			val[i] = 'K';
			i++;
		}
		else if (val[i] == 20) {
			val[i] = 'L';
			i++;
		}
		else if (val[i] == 15) {
			val[i] = 'M';
			i++;
		}
		else if (val[i] == 23) {
			val[i] ='N';
			i++;
		}
		else if (val[i] == 25) {
			val[i] = 'O';
			i++;
		}
		else if (val[i] == 8) {
			val[i] = 'P';
			i++;
		}
		else if (val[i] == 24) {
			val[i] = 'Q';
			i++;
		}
		else if (val[i] == 21) {
			val[i] = 'R';
			i++;
		}
		else if (val[i] == 19) {
			val[i] = 'S';
			i++;
		}
		else if (val[i] == 16) {
			val[i] = 'T';
			i++;
		}
		else if (val[i] == 1) {
			val[i] = 'U';
			i++;
		}
		else if (val[i] == 9) {
			val[i] = 'V';
			i++;
		}
		else if (val[i] == 2) {
			val[i] = 'W';
			i++;
		}
		else if (val[i] == 18) {
			val[i] = 'X';
			i++;
		}
		else if (val[i] == 3) {
			val[i] = 'Y';
			i++;
		}
		else if (val[i] == 10) {
			val[i] = 'Z';
			i++;
		}
		else if (val[i] == 32) {
			val[i] = 'a';
			i++;
		}
		else if (val[i] == 38) {
			val[i] = 'b';
			i++;
		}
		else if (val[i] == 40) {
			val[i] = 'c';
			i++;
		}
		else if (val[i] == 33) {
			val[i] = 'd';
			i++;
		}
		else if (val[i] == 39) {
			val[i] = 'e';
			i++;
		}
		else if (val[i] == 41) {
			val[i] = 'f';
			i++;
		}
		else if (val[i] == 44) {
			val[i] = 'g';
			i++;
		}
		else if (val[i] == 49) {
			val[i] = 'h';
			i++;
		}
		else if (val[i] == 53) {
			val[i] = 'i';
			i++;
		}
		else if (val[i] == 41) {
			val[i] = 'j';
			i++;
		}
		else if (val[i] == 47) {
			val[i] = 'k';
			i++;
		}
		else if (val[i] == 42) {
			val[i] =  'l';
			i++;
		}
		else if (val[i] == 50) {
			val[i] =  'm';
			i++;
		}
		else if (val[i] == 52) {
			val[i] = 'n';
			i++;
		}
		else if (val[i] == 35) {
			val[i] = 'o';
			i++;
		}
		else if (val[i] == 51) {
			val[i] = 'p';
			i++;
		}
		else if (val[i] == 48) {
			val[i] = 'q';
			i++;
		}
		else if (val[i] == 46) {
			val[i] = 'r';
			i++;
		}
		else if (val[i] == 43) {
			val[i] = 's';
			i++;
		}
		else if (val[i] == 28) {
			val[i] = 't';
			i++;
		}
		else if (val[i] == 36) {
			val[i] = 'u';
			i++;
		}
		else if (val[i] == 29) {
			val[i] = 'v';
			i++;
		}
		else if (val[i] == 45) {
			val[i] = 'w';
			i++;
		}
		else if (val[i] == 30) {
			val[i] = 'x';
			i++;
		}
		else if (val[i] == 37) {
			val[i] = 'y';
			i++;
		}
		else if (val[i] == 59) {
			val[i] = 'z';
			i++;
		}
		else if (val[i] == 65) {
			val[i] = '0';
			i++;
		}
		else if (val[i] == 67) {
			val[i] = '1';
			i++;
		}
		else if (val[i] == 60) {
			val[i] = '2';
			i++;
		}
		else if (val[i] == 66) {
			val[i] = '3';
			i++;
		}
		else if (val[i] == 58) {
			val[i] = '4';
			i++;
		}
		else if (val[i] == 71) {
			val[i] = '5';
			i++;
		}
		else if (val[i] == 76) {
			val[i] = '6';
			i++;
		}
		else if (val[i] == 80) {
			val[i] = '7';
			i++;
		}
		else if (val[i] == 68) {
			val[i] = '8';
			i++;
		}
		else if (val[i] == 74) {
			val[i] = '9';
			i++;
		}
		else if (val[i] == 69) {
			val[i] = '!';
			i++;
		}
		else if (val[i] == 77) {
			val[i] = '@';
			i++;
		}
		else if (val[i] == 79) {
			val[i] = '#';
			i++;
		}
		else if (val[i] == 62) {
			val[i] = '$';
			i++;
		}
		else if (val[i] == 78) {
			val[i] = '%';
			i++;
		}
		else if (val[i] == 75) {
			val[i] = '^';
			i++;
		}
		else if (val[i] == 73) {
			val[i] = '&';
			i++;
		}
		else if (val[i] == 70) {
			val[i] = '*';
			i++;
		}
		else if (val[i] == 55) {
			val[i] = '(';
			i++;
		}
		else if (val[i] == 63) {
			val[i] = ')';
			i++;
		}
		else if (val[i] == 56) {
			val[i] = '_';
			i++;
		}
		else if (val[i] == 72) {
			val[i] = '-';
			i++;
		}
		else if (val[i] == 57) {
			val[i] = '+';
			i++;
		}
		else if (val[i] == 64) {
			val[i] = '=';
			i++;
		}
		else if (val[i] == 86) {
			val[i] = '[';
			i++;
		}
		else if (val[i] == 92) {
			val[i] = '{';
			i++;
		}
		else if (val[i] == 94) {
			val[i] = ']';
			i++;
		}
		else if (val[i] == 87) {
			val[i] = '}';
			i++;
		}
		else if (val[i] == 93) {
			val[i] = 96;
			i++;
		}
		else if (val[i] == 85) {
			val[i] = 34;
			i++;
		}
		else if (val[i] == 98) {
			val[i] = 92;
			i++;
		}
		else if (val[i] == 103) {
			val[i] = '|';
			i++;
		}
		else if (val[i] == 107) {
			val[i] = '.';
			i++;
		}
		else if (val[i] == 95) {
			val[i] = '>';
			i++;
		}
		else if (val[i] == 101) {
			val[i] = '<';
			i++;
		}
		else if (val[i] == 96) {
			val[i] = ',';
			i++;
		}
		else if (val[i] == 104) {
			val[i] = ' ';
			i++;
		}
		else if (val[i] == 106) {
			val[i] = '/';
			i++;
		}
		else if (val[i] == 89) {
			val[i] = '\?';
			i++;
		}
		else if (val[i] == 105) {
			val[i] = '\a';
			i++;
		}
		else if (val[i] == 102) {
			val[i] = '\b';
			i++;
		}
		else if (val[i] == 100) {
		// escape character maybe not be supported if used as \e. the ascii is 27 and hex is 1B
			val[i] = 27;
			i++;
		}
		else if (val[i] == 97) {
			val[i] = '\f';
			i++;
		}
		else if (val[i] == 82){
			val[i] = '\n';
			i++;
		}
		else if (val[i] == 90){ 
			val[i] = '\r';
			i++;
		}
		else if (val[i] == 83) {
			val[i] = '\t';
			i++;
		}
		else if (val[i] == 99) {
			val[i] = '\v';
			i++;
		}
		else if (val[i] == 84) {
			val[i] = '\'';
			i++;
		}
		else{
			printf("%d, ", val[i]);
			i++;
		}
		/* not used yet
		else if (val[i] == 101) {
			val[i] = 5;
			i++;
		}
		else if (val[i] == 102) {
			val[i] = 6;
			i++;
		}
		else if (val[i] == 103) {
			val[i] = 4;
			i++;
		}
		else if (val[i] == 104) {
			val[i] = 17;
			i++;
		}
		else if (val[i] == 105) {
			val[i] = 22;
			i++;
		}
		else if (val[i] == 106) {
			val[i] = 26;
			i++;
		}
		else if (val[i] == 107) {
			val[i] = 14;
			i++;
		}
		else if (val[i] == 108) {
			val[i] = 20;
			i++;
		}
		else if (val[i] == 109) {
			val[i] = 15;
			i++;
		}
		else if (val[i] == 110) {
			val[i] = 23;
			i++;
		}
		else if (val[i] == 111) {
			val[i] = 25;
			i++;
		}
		else if (val[i] == 112) {
			val[i] = 24;
			i++;
		}
		else if (val[i] == 113) {
			val[i] = 21;
			i++;
		}
		else if (val[i] == 114) {
			val[i] = 19;
			i++;
		}
		else if (val[i] == 115) {
			val[i] = 16;
			i++;
		}
		else if (val[i] == 116) {
			val[i] = 1;
			i++;
		}
		else if (val[i] == 117) {
			val[i] = 2;
			i++;
		}
		else if (val[i] == 118) {
			val[i] = 18;
			i++;
		}
		else if (val[i] == 119) {
			val[i] = 3;
			i++;
		}
		else if (val[i] == 120) {
			val[i] = 28;
			i++;
		}
		else if (val[i] == 121) {
			val[i] = 29;
			i++;
		}
		else if (val[i] == 122) {
			val[i] = 30;
			i++;
		}
		else if (val[i] == 123) {
			val[i] = 59;
			i++;
		}
		else if (val[i] == 124) {
			val[i] = 58;
			i++;
		}
		else if (val[i] == 125) {
			val[i] = 64;
			i++;
		}
		*/
	}
}


void checkDecipherPhase()
{
	if (phase1 > 33 || phase2 > 33 || phase3 > 33) {
		printf("%s\n", "Error: Max value for decipher phase is 33");
		exit(EXIT_FAILURE);
	}
	else if (phase1 < 1 || phase2 < 1 || phase3 < 1) {
		printf("%s\n", "Error: Minimum value for decipher phase is 1. You can use value 0 but not recommended");
		exit(EXIT_FAILURE);
	}
}

void decipher(int *val, int len)
{
	// Your phase must be same as your initial phase you use to encrypt your content in order to decrypt it
	phase1 = 33;
	phase2 = 33;
	phase3 = 32;
	checkDecipherPhase();
	int i = 0;
	while (i < len) {
			val[i] = val[i] - phase1 - phase2 - phase3;
			// if value become negative then convert to positive number in order to output to a file
			// This prevent the program from going into hang mode when can't decrypt
			if (val[i] < 0) {
				val[i] = abs(val[i]);
			}
			add();
			i++;
		}
}

void decipher2ascii(int *val, int len)
{
	int i = 0;
	while (i < len) {
		if (val[i] == 1) {
			val[i] = 'a';
			i++;
		}
		else if (val[i] == 2) {
			val[i] = 'b';
			i++;
		}
		else if (val[i] == 3) {
			val[i] = 'c';
			i++;
		}
		else if (val[i] == 4) {
			val[i] = 'd';
			i++;
		}
		else if (val[i] == 5) {
			val[i] = 'e';
			i++;
		}
		else if (val[i] == 6) {
			val[i] = 'f';
			i++;
		}
		else if (val[i] == 7) {
			val[i] = 'g';
			i++;
		}
		else if (val[i] == 8) {
			val[i] = 'h';
			i++;
		}
		else if (val[i] == 9) {
			val[i] = 'i';
			i++;
		}
		else if (val[i] == 10) {
			val[i] = 'j';
			i++;
		}
		else if (val[i] == 11) {
			val[i] = 'k';
			i++;
		}
		else if (val[i] == 12) {
			val[i] = 'l';
			i++;
		}
		else if (val[i] == 13) {
			val[i] = 'm';
			i++;
		}
		else if (val[i] == 14) {
			val[i] ='n';
			i++;
		}
		else if (val[i] == 15) {
			val[i] = 'o';
			i++;
		}
		else if (val[i] == 16) {
			val[i] = 'p';
			i++;
		}
		else if (val[i] == 17) {
			val[i] = 'q';
			i++;
		}
		else if (val[i] == 18) {
			val[i] = 'r';
			i++;
		}
		else if (val[i] == 19) {
			val[i] = 's';
			i++;
		}
		else if (val[i] == 20) {
			val[i] = 't';
			i++;
		}
		else if (val[i] == 21) {
			val[i] = 'u';
			i++;
		}
		else if (val[i] == 22) {
			val[i] = 'v';
			i++;
		}
		else if (val[i] == 23) {
			val[i] = 'w';
			i++;
		}
		else if (val[i] == 24) {
			val[i] = 'x';
			i++;
		}
		else if (val[i] == 25) {
			val[i] = 'y';
			i++;
		}
		else if (val[i] == 26) {
			val[i] = 'z';
			i++;
		}
		else if (val[i] == 27) {
			val[i] = 'A';
			i++;
		}
		else if (val[i] == 28) {
			val[i] = 'B';
			i++;
		}
		else if (val[i] == 29) {
			val[i] = 'C';
			i++;
		}
		else if (val[i] == 30) {
			val[i] = 'D';
			i++;
		}
		else if (val[i] == 31) {
			val[i] = 'E';
			i++;
		}
		else if (val[i] == 32) {
			val[i] = 'F';
			i++;
		}
		else if (val[i] == 33) {
			val[i] = 'G';
			i++;
		}
		else if (val[i] == 34) {
			val[i] = 'H';
			i++;
		}
		else if (val[i] == 35) {
			val[i] = 'I';
			i++;
		}
		else if (val[i] == 36) {
			val[i] = 'J';
			i++;
		}
		else if (val[i] == 37) {
			val[i] = 'K';
			i++;
		}
		else if (val[i] == 38) {
			val[i] =  'L';
			i++;
		}
		else if (val[i] == 39) {
			val[i] =  'M';
			i++;
		}
		else if (val[i] == 40) {
			val[i] = 'N';
			i++;
		}
		else if (val[i] == 41) {
			val[i] = 'O';
			i++;
		}
		else if (val[i] == 42) {
			val[i] = 'P';
			i++;
		}
		else if (val[i] == 43) {
			val[i] = 'Q';
			i++;
		}
		else if (val[i] == 44) {
			val[i] = 'R';
			i++;
		}
		else if (val[i] == 45) {
			val[i] = 'S';
			i++;
		}
		else if (val[i] == 46) {
			val[i] = 'T';
			i++;
		}
		else if (val[i] == 47) {
			val[i] = 'U';
			i++;
		}
		else if (val[i] == 48) {
			val[i] = 'V';
			i++;
		}
		else if (val[i] == 49) {
			val[i] = 'W';
			i++;
		}
		else if (val[i] == 50) {
			val[i] = 'X';
			i++;
		}
		else if (val[i] == 51) {
			val[i] = 'Y';
			i++;
		}
		else if (val[i] == 52) {
			val[i] = 'Z';
			i++;
		}
		else if (val[i] == 53) {
			val[i] = '0';
			i++;
		}
		else if (val[i] == 54) {
			val[i] = '1';
			i++;
		}
		else if (val[i] == 55) {
			val[i] = '2';
			i++;
		}
		else if (val[i] == 56) {
			val[i] = '3';
			i++;
		}
		else if (val[i] == 57) {
			val[i] = '4';
			i++;
		}
		else if (val[i] == 58) {
			val[i] = '5';
			i++;
		}
		else if (val[i] == 59) {
			val[i] = '6';
			i++;
		}
		else if (val[i] == 60) {
			val[i] = '7';
			i++;
		}
		else if (val[i] == 61) {
			val[i] = '8';
			i++;
		}
		else if (val[i] == 62) {
		val[i] = '9';
			i++;
		}
		else if (val[i] == 63) {
			val[i] = '!';
			i++;
		}
		else if (val[i] == 64) {
			val[i] = '@';
			i++;
		}
		else if (val[i] == 65) {
			val[i] = '#';
			i++;
		}
		else if (val[i] == 66) {
			val[i] = '$';
			i++;
		}
		else if (val[i] == 67) {
			val[i] = '%';
			i++;
		}
		else if (val[i] == 68) {
			val[i] = '^';
			i++;
		}
		else if (val[i] == 69) {
			val[i] = '&';
			i++;
		}
		else if (val[i] == 70) {
			val[i] = '*';
			i++;
		}
		else if (val[i] == 71) {
			val[i] = '(';
			i++;
		}
		else if (val[i] == 72) {
			val[i] = ')';
			i++;
		}
		else if (val[i] == 73) {
			val[i] = '_';
			i++;
		}
		else if (val[i] == 74) {
			val[i] = '-';
			i++;
		}
		else if (val[i] == 75) {
			val[i] = '+';
			i++;
		}
		else if (val[i] == 76) {
			val[i] = '=';
			i++;
		}
		else if (val[i] == 77) {
			val[i] = '[';
			i++;
		}
		else if (val[i] == 78) {
			val[i] = '{';
			i++;
		}
		else if (val[i] == 79) {
			val[i] = ']';
			i++;
		}
		else if (val[i] == 80) {
			val[i] = '}';
			i++;
		}
		else if (val[i] == 81) {
			val[i] = 96;
			i++;
		}
		else if (val[i] == 82) {
			val[i] = 34;
			i++;
		}
		else if (val[i] == 83) {
			val[i] = 92;
			i++;
		}
		else if (val[i] == 84) {
			val[i] = '|';
			i++;
		}
		else if (val[i] == 85) {
			val[i] = '.';
			i++;
		}
		else if (val[i] == 86) {
			val[i] = '>';
			i++;
		}
		else if (val[i] == 87) {
			val[i] = '<';
			i++;
		}
		else if (val[i] == 88) {
			val[i] = ',';
			i++;
		}
		else if (val[i] == 89) {
			val[i] = ' ';
			i++;
		}
		else if (val[i] == 90) {
			val[i] = '/';
			i++;
		}
		else if (val[i] == 91) {
			val[i] = '\?';
			i++;
		}
		else if (val[i] == 92) {
			val[i] = '\a';
			i++;
		}
		else if (val[i] == 93) {
			val[i] = '\b';
			i++;
		}
		else if (val[i] == 94) {
			val[i] = '\e';
			i++;
		}
		else if (val[i] == 95) {
			val[i] = '\f';
			i++;
		}
		else if (val[i] == 96){
			val[i] = '\n';
			i++;
		}
		else if (val[i] == 97){ 
			val[i] = '\r';
			i++;
		}
		else if (val[i] == 98) {
			val[i] = '\t';
			i++;
		}
		else if (val[i] == 99) {
			val[i] = '\v';
			i++;
		}
		else if (val[i] == 100) {
			val[i] = '\'';
			i++;
		}
		else if (val[i] == 101) {
			val[i] = 5;
			i++;
		}
		else if (val[i] == 102) {
			val[i] = 6;
			i++;
		}
		else if (val[i] == 103) {
			val[i] = 4;
			i++;
		}
		else if (val[i] == 104) {
			val[i] = 17;
			i++;
		}
		else if (val[i] == 105) {
			val[i] = 22;
			i++;
		}
		else if (val[i] == 106) {
			val[i] = 26;
			i++;
		}
		else if (val[i] == 107) {
			val[i] = 14;
			i++;
		}
		else if (val[i] == 108) {
			val[i] = 20;
			i++;
		}
		else if (val[i] == 109) {
			val[i] = 15;
			i++;
		}
		else if (val[i] == 110) {
			val[i] = 23;
			i++;
		}
		else if (val[i] == 111) {
			val[i] = 25;
			i++;
		}
		else if (val[i] == 112) {
			val[i] = 24;
			i++;
		}
		else if (val[i] == 113) {
			val[i] = 21;
			i++;
		}
		else if (val[i] == 114) {
			val[i] = 19;
			i++;
		}
		else if (val[i] == 115) {
			val[i] = 16;
			i++;
		}
		else if (val[i] == 116) {
			val[i] = 1;
			i++;
		}
		else if (val[i] == 117) {
			val[i] = 2;
			i++;
		}
		else if (val[i] == 118) {
			val[i] = 18;
			i++;
		}
		else if (val[i] == 119) {
			val[i] = 3;
			i++;
		}
		else if (val[i] == 120) {
			val[i] = 28;
			i++;
		}
		else if (val[i] == 121) {
			val[i] = 29;
			i++;
		}
		else if (val[i] == 122) {
			val[i] = 30;
			i++;
		}
		else if (val[i] == 123) {
			val[i] = ';';
			i++;
		}
		else if (val[i] == 124) {
			val[i] = ':';
			i++;
		}
		else if (val[i] == 125) {
			val[i] = 64;
			i++;
		}

	}
}